arbor/communication/communicator.cpp

```diff
@@ -218,28 +241,45 @@ time_type communicator::min_delay() {
 }
 
 communicator::spikes
-communicator::exchange(std::vector<spike> local_spikes) {
+communicator::exchange(std::vector<std::vector<spike>>& local_spikes) {
     PE(communication:exchange:sort);
     // sort the spikes in ascending order of source gid
-    util::sort_by(local_spikes, [](spike s){return s.source;});
+    for (auto& rank : local_spikes) {
+        util::sort_by(rank, [](const spike& s) { return s.source; });
+    }
     PL();
 
-    PE(communication:exchange:gather);
-    // global all-to-all to gather a local copy of the global spike list on each node.
-    auto global_spikes = ctx_->distributed->gather_spikes(local_spikes);
+    PE(communication:exchange:all2all);
+    auto global_spikes = ctx_->distributed->all_to_all_spikes(local_spikes);
     num_spikes_ += global_spikes.size();
     PL();
 
     // Get remote spikes
     PE(communication:exchange:gather:remote);
     if (remote_spike_filter_) {
-        local_spikes.erase(std::remove_if(local_spikes.begin(),
-                                          local_spikes.end(),
-                                          [this] (const auto& s) { return !remote_spike_filter_(s); }));
+        for (auto& buffer : local_spikes) {
+            buffer.erase(std::remove_if(buffer.begin(), buffer.end(),
+                                        [this](const spike& s) {
+                                            return !remote_spike_filter_(s);
+                                        }),
+                         buffer.end());
+        }
+    }
+    
+    std::vector<spike> flat_spikes;
+    flat_spikes.reserve(
+        std::accumulate(local_spikes.begin(), local_spikes.end(), 0u,
+            [](std::size_t acc, const auto& v) { return acc + v.size(); }));
+
+    for (const auto& subvec : local_spikes) {
+        flat_spikes.insert(flat_spikes.end(), subvec.begin(), subvec.end());
     }
-    auto remote_spikes = ctx_->distributed->remote_gather_spikes(local_spikes);
-    PL();
 
+    std::sort(flat_spikes.begin(), flat_spikes.end()); 
+    flat_spikes.erase(std::unique(flat_spikes.begin(), flat_spikes.end()), flat_spikes.end());
+    
+    auto remote_spikes = ctx_->distributed->remote_gather_spikes(flat_spikes);
```

Comment:
This seems like one unfortunate consequence of the nested vector datastructure, especially as the overhead is payed even if no cosimulation partner exists.